src/osd/OSD.cc

```diff
@@ -7347,6 +7349,25 @@ void OSD::handle_osd_map(MOSDMap *m)
     skip_maps = true;
   }
 
+  // wait for pgs to catch up
+  {
+    // we extend the map cache pins to accomodate pgs slow to consume maps
+    // for some period, until we hit the max_lag_factor bound, at which point
+    // we block here to stop injesting more maps than they are able to keep
+    // up with.
+    epoch_t max_lag = cct->_conf->osd_map_cache_size *
+      m_osd_pg_epoch_max_lag_factor;
+    assert(max_lag > 0);
+    if (osdmap->get_epoch() > max_lag) {
+      epoch_t min = osdmap->get_epoch() - max_lag;
```

Comment:
So we're entering this block whenever the osdmap epoch is bigger than our max lag (so, always, after we have run for a little while). Are we just counting on the service.wait_min_pg_epoch() call to return instantly in those cases?
It does, other than grabbing the lock, so it works. The code structure and logging output led me to think we'd only do the call if we thought we'd exceeded our bounds, though; we shouldn't output at level 10 for a non-"failure" for instance...

Reply:
fixed this message to be more informative